### PR TITLE
Add string interface to updater to make it consistent with kvstore

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
example:
```
push(1, grad) -> updater.__call__(1, weight, g)
push('1', grad) -> updater.__call__('1', weight, g)
```

### CODE REVIEW

**Comment by eric-haibin-lin** [comment]
This change requires the frontend to call SetupUpdaterEx which sets the `_str_updater_` and `_updater_` and backend. If `_str_updater_` is missing when string key is used, `_updater_` will be called for backward compatibility.  

TODO:
add string key interface to updater for dist-kvstore as well. 

@reminisce @piiswrong @mli

File: python/mxnet/kvstore.py [hunk 1]
```diff
@@ -29,26 +29,37 @@
...
-            c_key_i, c_val_i = _ctype_key_value(key, val)
+            c_key_i, c_val_i, t = _ctype_key_value(key, val)
             c_keys += c_key_i
             c_vals += c_val_i
-        return (c_array(ctypes.c_char_p, c_keys), c_array(NDArrayHandle, c_vals))
-    names = []
-    keys = str(keys)
+            key_type = t if key_type is None else key_type
+            assert(key_type == t), "inconsistent types of keys detected."
+        c_keys_arr = c_array(ctypes.c_char_p, c_keys) if key_type == str \
```
Comment (piiswrong) [inline python/mxnet/kvstore.py]: use base.string_types instead of str

File: src/kvstore/kvstore_local.h [hunk 2]
```diff
@@ -84,30 +86,7 @@ class KVStoreLocal : public KVStore {
...
-        updater_(key, merged,  &local);
-      } else {
-        if (merged.storage_type() != local.storage_type()) {
-          local = merged.Copy(local.ctx());
-        } else {
-          local = merged;
-        }
-      }
-    }
+    Push_(keys, values, priority, false);
```
Comment (piiswrong) [inline src/kvstore/kvstore_local.h]: Only allow int or str for one kvstore. Cannot use both.

**Comment by eric-haibin-lin** [comment]
Done

**Pull request merged**